qadence_protocols/measurements/utils.py

```diff
@@ -0,0 +1,179 @@
+        op: Tuple of Primitive blocks or None.
+
+    Returns: A list of integers representing qubit indices.
+    """
+    blocks = getattr(pauli_term[0], "blocks", None)
+    blocks = blocks if blocks is not None else [pauli_term[0]]
+    indices = [
+        block.qubit_support[0] for block in blocks if (op is None) or (isinstance(block, type(op)))
+    ]
+    return indices
+
+
+def get_counts(samples: List, support: List[int]) -> List[Counter]:
+    """Marginalise the probability mass function to the support.
+
+    Args:
+        samples: List of samples against which expectation value is to be computed.
+        support: A list of integers representing qubit indices.
+
+    Returns: A List[Counter] of bit strings.
+    """
+    return [
+        reduce(
+            lambda x, y: x + y,
+            [Counter({"".join([k[i] for i in support]): sample[k]}) for k, v in sample.items()],
+        )
+        for sample in samples
+    ]
+
+
+def empirical_average(samples: List, support: List[int]) -> Tensor:
+    """Compute the empirical average.
+
+    Args:
+        samples: List of samples against which expectation value is to be computed.
+        support: A list of integers representing qubit indices.
+
+    Returns: A torch.Tensor of the empirical average.
+    """
+    PARITY = -1
+    counters = get_counts(samples, support)
+    n_shots = np.sum(list(counters[0].values()))
+    expectations = []
+    for counter in counters:
+        counter_exps = []
+        for bitstring, count in counter.items():
+            counter_exps.append(count * PARITY ** (np.sum([int(bit) for bit in bitstring])))
+        expectations.append(np.sum(counter_exps) / n_shots)
+    return torch.tensor(expectations)
+
+
+def pauli_z_expectation(
+    pauli_decomposition: List[Tuple[AbstractBlock, Basic]],
+    samples: List[Counter],
+) -> Tensor:
+    """Estimate total expectation value from samples by averaging all Pauli terms.
+
+    Args:
+        pauli_decomposition: A list of Pauli decomposed terms.
+        samples: Samples of bit string as Counters.
+
+    Returns: A torch.Tensor of the total expectation.
+    """
+
+    estimated_values = []
+    for pauli_term in pauli_decomposition:
+        support = get_qubit_indices_for_op(pauli_term)
+        estim_values = empirical_average(samples=samples, support=support)
+        # TODO: support for parametric observables to be tested
+        estimated_values.append(estim_values * evaluate(pauli_term[1]))
+    res = torch.sum(torch.stack(estimated_values), axis=0)
+    return res
+
+
+def rotate(circuit: QuantumCircuit, pauli_term: Tuple[AbstractBlock, Basic]) -> QuantumCircuit:
+    """Rotate circuit to measurement basis and return the qubit support.
+
+    Args:
+        circuit: The circuit that is executed.
+        pauli_term: Tuple of a Pauli term and a parameter.
+
+    Returns: Rotated QuantumCircuit.
+    """
+
+    rotations = []
+
+    for op, gate in [(X(0), Z), (Y(0), SDagger)]:
```

Comment:
I never found out a nice solution for checking `X` and `Y` types other than instantiate them. Maybe there is something to do here.

Reply:
@Roland-djee I managed to remove the instantiations